File: app_flutter/app_cua_toi/lib/services/address_service.dart
import 'dart:async';
import 'package:dio/dio.dart';
import '../data/network/api_client.dart';
import '../data/models/api_response.dart';
import '../data/models/user_model.dart';

class AddressService {
  static final Dio _dio = ApiClient.instance;

  static final StreamController<List<Address>> _addressController =
      StreamController<List<Address>>.broadcast();
  static Stream<List<Address>> get addressStream => _addressController.stream;

  static List<Address> _currentAddresses = [];

  static Future<void> fetchAddresses(int customerId) async {
    try {
      final response = await _dio.get('/addresses/$customerId');
      if (response.data['code'] == 1000) {
        final list = response.data['result'] as List;
        _currentAddresses = list.map((e) => Address.fromJson(e)).toList();
        _addressController.add(_currentAddresses);
      }
    } catch (e) {
      print('Fetch Address Error: $e');
    }
  }

  static Future<ApiResponse<List<Address>>> getAddressesByCustomerId(
    int customerId,
  ) async {
    try {
      final response = await _dio.get('/addresses/$customerId');
      final apiResponse = ApiResponse<List<Address>>.fromJson(response.data, (
        json,
      ) {
        final list = json as List;
        return list.map((e) => Address.fromJson(e)).toList();
      });

      if (apiResponse.code == 1000) {
        _currentAddresses = apiResponse.result ?? [];
        _addressController.add(_currentAddresses);
      }
      return apiResponse;
    } catch (e) {
      return _handleError(e);
    }
  }

  static Future<ApiResponse<Address>> createAddress({
    required int customerId,
    required String address,
    required String province,
    required String district,
    required String ward,
  }) async {
    try {
      final response = await _dio.post(
        '/addresses',
        data: {
          'customerId': customerId,
          'address': address,
          'province': province,
          'district': district,
          'ward': ward,
        },
      );

      final apiResponse = ApiResponse<Address>.fromJson(
        response.data,
        (json) => Address.fromJson(json),
      );

      if (apiResponse.code == 1000) {
        await fetchAddresses(customerId); // Refresh list
      }
      return apiResponse;
    } catch (e) {
      return _handleError(e);
    }
  }

  static Future<ApiResponse<String>> deleteAddress(
    int addressId,
    int customerId,
  ) async {
    try {
      final response = await _dio.delete('/addresses/$addressId');
      final apiResponse = ApiResponse<String>.fromJson(
        response.data,
        (json) => json.toString(),
      );

      if (apiResponse.code == 1000) {
        await fetchAddresses(customerId); // Refresh list
      }
      return apiResponse;
    } catch (e) {
      return _handleError(e);
    }
  }

  static ApiResponse<T> _handleError<T>(dynamic e) {
    String msg = "Lỗi kết nối";
    int code = 9999;
    if (e is DioException && e.response != null) {
      final data = e.response!.data;
      if (data is Map) {
        msg = data['message'] ?? msg;
        code = data['code'] ?? code;
      }
    }
    return ApiResponse<T>(code: code, message: msg);
  }

  static void dispose() {
    _currentAddresses = [];
    _addressController.add([]);
  }
}

File: giao_dien_quan_ly/webapp/src/api/services/address.service.ts
import axiosClient from "../axiosClient";
import { ApiResponse } from "../types/api.response";
import { AddressRequest, AddressResponse } from "../types/customer.types";

const BASE_URL = "/addresses";

export const addressService = {
  // GET /addresses/{customerId} (Lấy danh sách địa chỉ theo ID khách hàng)
  getByCustomerId: async (customerId: number) => {
    const response = await axiosClient.get<any, ApiResponse<AddressResponse[]>>(
      `${BASE_URL}/${customerId}`
    );
    return response.result!;
  },

  // POST /addresses (Tạo mới địa chỉ)
  create: async (data: AddressRequest & { customerId: number }) => {
    const response = await axiosClient.post<any, ApiResponse<AddressResponse>>(
      BASE_URL,
      data
    );
    return response.result!;
  },

  // DELETE /addresses/{addressId} (Xóa địa chỉ)
  delete: async (addressId: number) => {
    const response = await axiosClient.delete<any, ApiResponse<string>>(
      `${BASE_URL}/${addressId}`
    );
    return response.result;
  },
};